dali/kernels/reduce/reduce_setup_utils.h

```diff
@@ -0,0 +1,229 @@
+ *                   first dimension in group and count is the number of collapsed dimensions.
+ *
+ * @param in_shape   shape of the input
+ * @param axes       list of reduced axes; axis indices must be <= 63
+ *
+ * A dimension can be collapsed with its neighbor if:
+ *  - its extent is 1 in all samples
+ *  - both dimensions are reduced or non-reduced.
+ */
+template <typename Axes, typename DimGroups, int ndim>
+void SimplifyReduction(Axes &out_axes, DimGroups &dim_groups,
+                       const TensorListShape<ndim> &in_shape, span<const int> axes) {
+  out_axes.clear();
+  dim_groups.clear();
+  int d = in_shape.sample_dim();
+
+  uint64_t mask = to_bit_mask(axes);
+
+  int i;
+
+  // skip leading degenerate dimensions
+  for (i = 0; i < d && is_degenerate_dim(in_shape, i); i++) {}
+
+  if (i >= d) {
+    // no non-degenerate dimensions
+    dim_groups.push_back(std::make_pair(0, d));
+    return;
+  }
+
+  // so we've reached the outermost non-degenerate dimension
+  int remapped = 0;  // its index will map to 0
+
+  bool prev_reduced = (mask >> i) & 1;  // is it reduced?
+  if (prev_reduced)
+    out_axes.push_back(remapped);  // if so, add the axis to the list of reduced dimension groups
+
+  int group_start = 0;     // the dimension group starts at 0 -
+  int group_size = i + 1;  // all leading degenerate dimensions are in this group
+
+  for (i++; i < d; i++) {
+    if (is_degenerate_dim(in_shape, i)) {
+      // Degenerate dimension can be merged with the previous one regardless of whether
+      // it is reduced or not.
+      group_size++;
+      continue;
+    }
+    bool is_reduced = (mask >> i) & 1;
+    if (is_reduced != prev_reduced) {
+      dim_groups.emplace_back(group_start, group_size);
+      group_start = i;
+      group_size = 1;
+      remapped++;
+      if (is_reduced)
+        out_axes.push_back(remapped);
+    } else {
+      group_size++;
+    }
+    prev_reduced = is_reduced;
+  }
+  dim_groups.push_back(std::make_pair(group_start, group_size));
+}
+
+
+/**
+ * @brief Throws an exception if given tensor list cannot be reduced across samples with
+ *        given set of axes.
+ *
+ * A tensor list can be reduced across samples only if the non-reduced dimensions have the
+ * same extent in all samples.
+ */
+inline void CheckBatchReduce(const TensorListShape<> &tls, span<const int> axes) {
+  if (tls.num_samples() == 0)
+    return;
+
+  uint64_t mask = to_bit_mask(axes);
+  SmallVector<int, DynamicTensorShapeContainer::static_size> non_reduced;
+  for (int a = 0; a < tls.sample_dim(); a++) {
+    if (!(mask & (1 << a)))
+      non_reduced.push_back(a);
+  }
+
+  auto first_sample_shape = tls.tensor_shape_span(0);
+
+  for (int i = 1; i < tls.num_samples(); i++) {
+    auto sample_shape = tls.tensor_shape_span(i);
+    for (int a : non_reduced) {
+      DALI_ENFORCE(sample_shape[a] == first_sample_shape[a], make_string(
```

Comment:
DO you want to use `DALI_ENFORCE` inside kernel? I think the plain exception is the way.

Reply:
I'll replace it with appropriate error (invalid argument or logic error).